File: 2020/day/1/main.py
import fileinput

with fileinput.input('input.txt') as f:
    numbers = {int(ln) for ln in f}

# Part 1
for x in numbers:
    if 2020 - x in numbers:
        print(x * (2020 - x))
        break

# Part 2
done = False

for x in numbers:
    for y in numbers:
        if 2020 - x - y in numbers:
            print(x * y * (2020 - x - y))
            done = True
            break
    
    if done: break

File: 2020/day/1/Main.hs
import qualified Data.IntSet as Set

main = do
    input <- readFile "input.txt"
    let numbers = Set.fromList $ map read $ lines input

    -- Part 1
    print $ head [ x * (2020 - x)
                 | x <- Set.toList numbers
                 , (2020 - x) `Set.member` numbers
                 ]

    -- Part 2
    print $ head [ x * y * (2020 - x - y)
                 | x <- Set.toList numbers
                 , y <- Set.toList numbers
                 , (2020 - x - y) `Set.member` numbers
                 ]